floris/tools/wind_data.py

```diff
@@ -0,0 +1,529 @@
+# Copyright 2024 NREL
+
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# See https://floris.readthedocs.io for documentation
+
+import matplotlib.cm as cm
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from pandas.api.types import CategoricalDtype
+
+
+# Define the super lass that WindRose and TimeSeries inherit
+# Define functions here that are either the same for both WindRose and
+# TimeSeries or will be overloaded
+class WindDataBase:
+    def __init__():
+        pass
+
+    def unpack_for_reinitialize(self):
+        """
+        Return only the variables need for reinitialize
+        """
+        (
+            wind_directions_unpack,
+            wind_speeds_unpack,
+            _,
+            ti_table_unpack,
+            _,
+        ) = self.unpack()
+
+        return wind_directions_unpack, wind_speeds_unpack, ti_table_unpack
+
+    def unpack_freq(self):
+        """Unpack frequency weighting"""
+
+        (
+            _,
+            _,
+            freq_table_unpack,
+            _,
+            _,
+        ) = self.unpack()
+
+        return freq_table_unpack
+
+
+class WindRose(WindDataBase):
+    """
+    In FLORIS v4, the WindRose class is used to drive FLORIS and optimization
+    operations in which the inflow is characterized by the frequency of
+    binned wind speed, wind direction and turbulence intensity values
+
+    Args:
+        wind_directions: NumPy array of wind directions (NDArrayFloat).
+        wind_speeds: NumPy array of wind speeds (NDArrayFloat).
+        freq_table: Frequency table for binned wind direction, wind speed
+            values (NDArrayFloat, optional).  Defaults to None.
+        ti_table: Turbulence intensity table for binned wind direction, wind
+            speed values (NDArrayFloat, optional).  Defaults to None.
+        price_table: Price table for binned binned wind direction, wind
+            speed values (NDArrayFloat, optional).  Defaults to None.
+        compute_zero_freq_occurrence: Flag indicating whether to compute zero
+            frequency occurrences (bool, optional).  Defaults to False.
+
+    """
+
+    def __init__(
+        self,
+        wind_directions,
+        wind_speeds,
+        freq_table=None,
+        ti_table=None,
+        price_table=None,
+        compute_zero_freq_occurence=False,
+    ):
+        if not isinstance(wind_directions, np.ndarray):
```

Comment:
Since you're enforcing the type, add type hints so users can easily know the function signature.

Reply:
isn't NDArrayFloat the type hint?

Reply:
Gotcha, so this comment is good then, my bad